py/src/nupic/optimization/ae.py

```diff
@@ -0,0 +1,898 @@
+    Convert a set of parameters into the data types used to represent them.
+    Returned result has the same structure as the parameters.
+    """
+    # Recurse through the parameter data structure.
+    if isinstance(default_parameters, dict):
+        return {key: _recursive_parameter_structure(value)
+            for key, value in default_parameters.items()}
+    elif isinstance(default_parameters, tuple):
+        return tuple(_recursive_parameter_structure(value)
+            for value in default_parameters)
+    # Determine data type of each entry in parameter data structure.
+    elif isinstance(default_parameters, float):
+        return float
+    elif isinstance(default_parameters, int):
+        return int
+    raise TypeError('Unaccepted type in experiment parameters: type "%s".'%(type(default_parameters).__name__))
+
+def _recursive_typecast_parameters(values, structure):
+    # Recurse through the parameter data structure.
+    if isinstance(structure, dict):
+        for key in structure:
+            values[key] = _recursive_typecast_parameters(values[key], structure[key])
+        return values
+    elif isinstance(structure, tuple):
+        return tuple(_recursive_typecast_parameters(*args)
+            for args in zip(values, structure))
+    # Type cast values.
+    elif structure == float:
+        value = float(values)
+        return float(str(value))
+    elif structure == int:
+        return int(round(float(values)))
+
+def paths(structure):
+    retval = []
+    if isinstance(structure, dict):
+        for key, value in structure.items():
+            retval.extend( "['%s']%s"%(key, path) for path in paths(value) )
+    elif isinstance(structure, tuple):
+        for idx, value in enumerate(structure):
+            retval.extend( "[%d]%s"%(idx, path) for path in paths(value) )
+    else:
+        retval.append('')
+    return sorted(retval)
+
+################################################################################
+
+
+def evaluate_default_parameters(lab):
+    # print('%X'%hash(lab.default_parameters))
+    return lab.default_parameters
+
+
+class EvaluateHashes:
+    def __init__(self, hashes):
+        self.hashes = [int(h, base=16) for h in hashes]
+
+    def __call__(self, lab):
+        try:
+            experiments = [lab.experiment_ids[h] for h in self.hashes]
+        except KeyError:
+            unknown = [h for h in self.hashes if h not in lab.experiment_ids]
+            raise ValueError('Hash not recognized: %X'%unknown[0])
+        rnd = random.random
+        return min(experiments, key=lambda x: x.attempts + rnd()).parameters
+        return random.choice(experiments).parameters
+
+
+def evaluate_all(lab):
+    rnd = random.random
+    return min(lab.experiments, key=lambda x: x.attempts + rnd()).parameters
+
+
+def evaluate_best(lab):
+    rnd = random.random
+    best = max(lab.experiments, key = lambda X: X.mean() + rnd() )
+    return best.parameters
+
+
+class GridSearch(object):
```

Comment:
do you need to write the hyper param optimization code yourself? It is available in many toolkits? 
https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html